### DISCUSSION TITLE
Directus as headless E-Commerce Backend

### DISCUSSION THREAD

**Discussion by madc** [directus/directus#15892] [Archive]
We've been using Directus mostly as Headless CMS and backend for Mobile Apps.
But I'm currently in the early stage of evaluating its use as a headless backend for a small/midsize e-commerce site.

There are some Github projects, but nothing mature, so I'm wondering if anyone in the Directus community has experiences in this field.
Additionally I'd also be glad to hear about any ideas, opinions and maybe pitfalls regarding this topic - everything is highly appreciated!

Thanks!

**Comment by rijkvanzanten** [comment]
Heya! As far as e-commerce goes, products/orders/invoices/etc are "just" data points in collections, so you should get most of the way there by making a couple collections, for example:

- products
- product_variants
- orders
- customers

We currently don't have any pre-made of the shelf starter guides for these use cases, but that's certainly something we want to get into in the future 🙂  (cc @erondpowell)

**Reply by erondpowell** [reply]
> that's certainly something we want to get into in the future

Yeah! As soon as we get a few more cornerstones of documentation down, these kinds of tutorials are going to be a big push.
@madc  Can you tell us more about your project and tech stack? I'm curious 🤠

**Reply by emrahnazif** [reply]
> * products
> * product_variants
> * orders
> * customers

What's the purpose of "customers" collection here? Why not to use the directus users?

**Reply by rijkvanzanten** [reply]
No real purpose, I was just thinking out loud about what types of data you'd need in a e-commerce setting. You're right @emrahnazif, you might as well use directus_users for your customers so you get all the access control built-in 🙂

**Reply by madc** [reply]
@erondpowell We started focusing on fully static websites a couple of years ago using Directus as a backend since the pre-release of version 9. Our frontend Stack currently is basically Vue/Nuxt (version 2, fully static). We rebuild the whore website on change fetching data from Directus directly via the GraphQL API.

As for our project, its a mid-sized webshop and timeline is not to strict. That's why we'd like to use it as a like to expand our modern stack to e-commerce and ditch all the old systems.

We'd want to use a hybrid approach: Generate static content wherever it makes sense and fetch everything we have to dynamically from a backend. For this backend we'd prefer a NodeJS based opensource project, as it's easier for us to extend/work with. as @rijkvanzanten pointed out, Directus enables us to create some basics (Customers Products,...), but I'm a bit worried about the processes that need to run server side (i.e. generate a new order). Personally I'd really love to use Directus for this, but I'm not sure about how to estimate the real effort. I'm currently in the process of mapping out the missing parts and will share them here.

@rijkvanzanten, @erondpowell: Both of you have expressed interest in exploring Directus as an e-commerce backend the future. Would you be generally interested in a cooperation to solve this?

**Reply by erondpowell** [reply]
> Both of you have expressed interest in exploring Directus as an e-commerce backend the future. Would you be generally interested in a cooperation to solve this?

btw Directus *is currently being used as an ecommerce backed* out in the wild. 
We just don't have an official guide to handle that atm.

@rijkvanzanten pings me *(I'm eron, the technical writer 🤓 good to meet you)* so I can read through and find out what users *are looking for* in Docs Land... That said, I'd love to hop on a call at some point to hear more about your project and then talk about your experience learning/using Directus. Getting to know the end-users (like you) really helps us dial-in our documentation and guides. Let me know if interested 🙏

**Reply by madc** [reply]
Hey @erondpowell, I'd gladly hop on a call with you (if this invitation was directed at me..)

Are there any details, you can share publicly about the existing e-commerce projects?
Potential interesting information could be:
- Size of the projects
- Are those projects highly specialized applications or "normal" webshops?
- How are background processes are tackled (extension vs. flow vs. external service)
- ...

**Reply by erondpowell** [reply]
Cool! Let's setup a time. Are you in the Discord group?
You can DM me there to setup a time. Otherwise, email me at eron@directus.io

**Good Questions:**
- In general, there are some pretty huge projects running out there, but I'm not 100% sure the details regarding ecommerce.
- You can just go in and configure your relational data model as you please. If there's an "edge-case feature" you require. Well, that's why its open-source, modular and extensible!
- When you say "background processes", can you give an example?

CC @directus/team for more details on "big e-commerce" and "background processes".

**Reply by madc** [reply]
We would not be interested in Directus it it wasn't for its openness and the amazing team behind it! You guys really do a great job! (this had to be said ;) ). The reason I'm asking these questions is, because we don't want to re-invent the wheel if we don't have to! And if we do, I'd love to contribute back to the community. 

"background processes" was a bad choice of words, I guess what I meant was server-side tasks. The most basic example I can think of is order-placement. When a user places an order, you want to double check everything server-side to prevent foul-play. With Directus I can see multiple ways of solving this, each one with up- and downsides of course. If there already is some kind of best practice, it would be very valuable.

**Reply by sshvaiko** [reply]
Hi @madc, did you make a choice or probably hear more from the Directus team?

Currently, I’m building an e-commerce + magazine website, and I try to choose between two options:
- Using Directus and extensions to build custom endpoints, cron tasks, integrations (Stripe, for example);
- Using Directus as ad admin panel for the content and NestJS for the backend.

In the second option I would work with data using Directus graphQL client and return it from NestJS to the front end. 

@erondpowell could you recommend anything for this case?

Thanks a lot!

**Reply by madc** [reply]
No, we haven't. Project start was moved back, which gives us more time for the technical concept.
Currently we've researched self hosted modern e-commerce systems. Based on NodeJS, we could not find a single one that is currently usable or likely to still be around in a couple of years. We had a look at [Saleor](https://saleor.io/), which seams generally stable but is written in Python and very resource-hungry.

So our current options are:
- Extend Directus with the functionality we need or
- use something we used before (likely Sylius, which is php based)

Please keep us posted on what your decisions are!

**Reply by sshvaiko** [reply]
Most likely it will be Directus + [Next.js](https://nextjs.org/), and additional logic will be implemented in Next.js using getServerSideProps and API Routes. I don't really think we need something like NestJS or any other framework for the backend as Directus provides great tools to work with data directly from Next.js.

So the idea is to simplify the solution by leveraging fewer tools :)

**Reply by erondpowell** [reply]
@sshvaiko 
> Using Directus and extensions to build custom endpoints, cron tasks, integrations (Stripe, for example)

> it will be Directus + [Next.js](https://nextjs.org/), and additional logic will be implemented in Next.js

based on the description, you can likely do all that functionality with [flows](https://docs.directus.io/reference/system/flows.html) as opposed to creating custom endpoints/extensions.
There are some wildly complex flows out there in the wild, some with dozens of operations.
And Projects with 100+ flows.

What logic are you looking to add into Next.js?

**Reply by sshvaiko** [reply]
@erondpowell Thanks for your answer! There will be registration and registered users should be able to add their own products. From your perspective, what will solve it better? Directus Flows/API Endpoints or Next.js?

**Reply by erondpowell** [reply]
>There will be registration and registered users should be able to add their own products. From your perspective, what will solve it better? Directus Flows/API Endpoints or Next.js?

It depends on implementation details

To start, the first statements sounds like you want:
- user registration and creation from the public role, so people can sign up themselves. (CC @freekrai )
- a `seller` role and a `buyer` role.
- custom access permissions for each `seller` and `buyer`

Checkout [users roles and permissions in Directus](https://docs.directus.io/configuration/users-roles-permissions.html) to learn more about granular access permissions.

Basically, you can enable those three things in Directus, regardless of the frontend you use.

The permissions configured will be applied to your API/SDK calls. If the API call is from public, it is granted pbulic permissions. it from `buyer` it has buyer permissions, etc.

---

> Directus Flows/API Endpoints or Next.js?

Flows provides task automation and data processing.
You could also use a framework like Next.js for a fair amount of task automation.
The choice totally depends on the specifics of your project.

--- 

Does that help?

**Reply by sshvaiko** [reply]
Yes, thanks @erondpowell! I will just start using Directus and implement all the logic inside.

**Reply by madc** [reply]
@sshvaiko This is me being curious again: Are you planning on using flows again?

On a side note, we "only" used flows to trigger deployment  via CI/CD and process contact forms and it is amazing! This feature blew my mind when i saw the first screenshot.

**Comment by erondpowell** [comment]
@madc 

> When a user places an order, you want to double check everything server-side to prevent foul-play.

Seems like you mean server-side logic, functions, validation. For that specific example, what do you want to add? like field-level regex validations? or something more robust? I'm wondering if it can all be done in a flow. I can imagine if you have 200 lines of validation logic, Flows > Run Script (currently, as of v9.18.x) might not be the best UX for that.

> Extend Directus with the functionality we need

Just got your email. Can't wait to hear more about that on our call.

> Based on NodeJS, we could not find a single one that is currently usable or likely to still be around in a couple of years. 

Wow. If we have time, I'd like to hear about your research process.

**Reply by fitu86** [reply]
Que tal, yo lo uso para gestionar cuentas en un marketplace de LATAM.

![Screenshot_20230402_165423_com android chrome~3](https://user-images.githubusercontent.com/16139127/229383378-e9d13d1a-f77c-4d1a-a70e-5881c734a3c1.jpg)
![Screenshot_20230402_170101_com android chrome~2](https://user-images.githubusercontent.com/16139127/229383380-a7c7f34a-757a-4619-9dc2-a80428fdf04e.jpg)
![Screenshot_20230401_134103](https://user-images.githubusercontent.com/16139127/229383394-02a6bc2c-fdb6-4c22-a7bd-89e9fe8d6440.jpg)

**Reply by eudigitalis** [reply]
> Que tal, yo lo uso para gestionar cuentas en un marketplace de LATAM.
> 
> ![Screenshot_20230402_165423_com android chrome~3](https://user-images.githubusercontent.com/16139127/229383378-e9d13d1a-f77c-4d1a-a70e-5881c734a3c1.jpg) ![Screenshot_20230402_170101_com android chrome~2](https://user-images.githubusercontent.com/16139127/229383380-a7c7f34a-757a-4619-9dc2-a80428fdf04e.jpg) ![Screenshot_20230401_134103](https://user-images.githubusercontent.com/16139127/229383394-02a6bc2c-fdb6-4c22-a7bd-89e9fe8d6440.jpg)

I'm a seller on MercadoLivre here in Brazil and I want to use Directus to create a PIM and manage two accounts I have on the marketplace.
Can you help me how you arrived at this result?

**Reply by fitu86** [reply]
I'm also a seller of auto parts in México. I use Directus to integrate with the MercadoLibre API, receiving notifications and querying resources and events from the accounts I manage. This application also integrates with a PIM and an ERP, both developed in Directus. I'd be happy to help you. I'd love to chat and share experiences with another seller who shares the same interest in Directus."
